moquette-0.16/broker/src/main/java/io/moquette/interception/BrokerInterceptor.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.moquette.interception;
+
+import io.moquette.BrokerConstants;
+import io.moquette.interception.messages.*;
+import io.moquette.broker.config.IConfig;
+import io.moquette.broker.subscriptions.Subscription;
+import io.netty.handler.codec.mqtt.MqttConnectMessage;
+import io.netty.handler.codec.mqtt.MqttPublishMessage;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import static io.moquette.logging.LoggingUtils.getInterceptorIds;
+
+/**
+ * An interceptor that execute the interception tasks asynchronously.
+ */
+public final class BrokerInterceptor implements Interceptor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BrokerInterceptor.class);
+    private final Map<Class<?>, List<InterceptHandler>> handlers;
+    private final ExecutorService executor;
+
+    private BrokerInterceptor(int poolSize, List<InterceptHandler> handlers) {
+        LOG.info("Initializing broker interceptor. InterceptorIds={}", getInterceptorIds(handlers));
+        this.handlers = new HashMap<>();
+        for (Class<?> messageType : InterceptHandler.ALL_MESSAGE_TYPES) {
+            this.handlers.put(messageType, new CopyOnWriteArrayList<InterceptHandler>());
+        }
+        for (InterceptHandler handler : handlers) {
+            this.addInterceptHandler(handler);
+        }
+        executor = Executors.newFixedThreadPool(poolSize);
+    }
+
+    /**
+     * Configures a broker interceptor, with a thread pool of one thread.
+     *
+     * @param handlers InterceptHandlers listeners.
+     */
+    public BrokerInterceptor(List<InterceptHandler> handlers) {
+        this(1, handlers);
+    }
+
+    /**
+     * Configures a broker interceptor using the pool size specified in the IConfig argument.
+     * @param props configuration properties.
+     * @param handlers InterceptHandlers listeners.
+     *
+     */
+    public BrokerInterceptor(IConfig props, List<InterceptHandler> handlers) {
+        this(Integer.parseInt(props.getProperty(BrokerConstants.BROKER_INTERCEPTOR_THREAD_POOL_SIZE, "1")), handlers);
+    }
+
+    /**
+     * Shutdown graciously the executor service
+     */
+    public void stop() {
+        LOG.info("Shutting down interceptor thread pool...");
+        executor.shutdown();
+        try {
+            LOG.info("Waiting for thread pool tasks to terminate...");
+            executor.awaitTermination(10L, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+        }
+        if (!executor.isTerminated()) {
+            LOG.warn("Forcing shutdown of interceptor thread pool...");
+            executor.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)